### PR TITLE
Add params and type hints to stop() #6075

### PR DESCRIPTION
**Pull request by shalearkane**
Fixes issue #6075 

### Overview
https://github.com/rucio/rucio/blob/f9d52b7c7afbc1cfa41076f1d24f85299aa06cb8/lib/rucio/daemons/conveyor/preparer.py#L42-L47 should accept parameters `signalnum` and `handler` before being used as as signal handler in https://github.com/rucio/rucio/blob/f9d52b7c7afbc1cfa41076f1d24f85299aa06cb8/bin/rucio-conveyor-preparer#L23-L28 according to the [python docs](https://docs.python.org/3/library/signal.html#signal.signal). However the same function is being called without parameters in Line 28 https://github.com/rucio/rucio/blob/f9d52b7c7afbc1cfa41076f1d24f85299aa06cb8/bin/rucio-conveyor-preparer#L28 So I added `*args` to the function parameters to resolve the issue. 

### Alternative Solution
The other possible solution is to add the parameters:
```python
def stop(signalnum, handler):
```
However, then for the function call in Line 28,
https://github.com/rucio/rucio/blob/f9d52b7c7afbc1cfa41076f1d24f85299aa06cb8/bin/rucio-conveyor-preparer#L28 I will have to pass those arguments to `stop()`.

### CODE REVIEW

**Review by dchristidis — CHANGES_REQUESTED** [review]
I think your choosing to use variadic arguments is reasonable.

This change should be applied to all daemons and not just the Conveyor Preparer. In addition, I think it’s a good opportunity to add some type hints, given that they are trivial.

**Comment by shalearkane** [comment]
On looking into other stop functions, I saw that they had parameters `signum` and `frame` which defaulted to `None`. So I changed the variadic argument to the same and added type hints.

File: lib/rucio/daemons/hermes/hermes.py [hunk 1]
```diff
@@ -337,7 +338,7 @@ def run_once_messages(heartbeat_handler: HeartbeatHandler,
     return must_sleep
 
 
-def stop(signum: int = None, frame: str = None) -> None:
```
Comment (shalearkane) [inline lib/rucio/daemons/hermes/hermes.py]: Here the type of `frame` is set to `str`. But according to this [discussion](https://github.com/python/typing/discussions/1042), it should be `FrameType`. Also the syntax of the type-hint is wrong.

Comment (dchristidis) [reply lib/rucio/daemons/hermes/hermes.py]: Which is why I found your choice of `*args` a nice compromise. The parameters aren’t used in the body of the function, so there is little to gain by being overly pedantic.

**Comment by shalearkane** [comment]
I have changed the arguments of the stop functions of other daemons to `*args`. I added type-hint `None` to the return type. Should I add type hints to other functions as well?

**Comment by dchristidis** [comment]
I was surprised to see that so many `stop()` functions have some parameters already. I’m now not as confident about the best approach and would welcome someone else to comment. They way I see it, there are two alternatives:

 1. `def stop(*args: Any) -> None:` It’s simple and, at least in my eyes, somewhat suggests that the arguments are ignored.
 2. `def stop(signum: Optional[int] = None, frame: Optional[FrameType] = None) -> None:` More complex, but communicates well that this function is a signal handler.

**Comment by bari12** [comment]
I prefer the second;

**Review by dchristidis — CHANGES_REQUESTED** [review]
Be sure to import `Optional` and `FrameType`. Frequently we do it like so:

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from types import FrameType
    from typing import Optional
```

And then place the type hints inside strings.

Please note that the commits should be squashed together. Finally, the commit message has some room for improvement.

**Comment by shalearkane** [comment]
I have imported the Types and I have fixed imports in several files and squashed the commit. Please review the commit message. Should I make it more descriptive?

**Review by dchristidis — CHANGES_REQUESTED** [review]
> And then place the type hints inside strings.

You missed this part. Type hints are evaluated and will lead to `NameError` exceptions if the necessary classes are not imported at runtime.

As for the commit message:
 1. It should follow the format `<component>: <change_message> #<issue number>`. You did well, but since you are touching multiple components, you should omit the first part (as opposed to prefixing the message with ‘bug’).
 2. The rest of the message can be misleading. ‘add default arguments’ suggests that the parameters already exist, which is not the case. As such, ‘add parameters’ is more appropriate.
 3. The body of the message is empty. You should use it to document *why* you are making this change. Explain how this is the proper signature for signal handlers and how deviating from it can lead to a `TypeError`.

**Comment by shalearkane** [comment]
I have updated my patch against the latest master branch and amended my commit message accordingly. Please review my changes :smile:

**Review by dchristidis — CHANGES_REQUESTED** [review]
Please review the failed tests. It would appear that there are some missing imports.

For example:
`lib/rucio/daemons/conveyor/preparer.py:42:49: F821 undefined name 'FrameType'`

Similarly, in some places you moved the import statement under a conditional block (i.e. `if TYPE_CHECKING`) for things that are already in use.

Finally, you’ll need to rebase your branch on the latest master.

File: lib/rucio/daemons/atropos/atropos.py [hunk 2]
```diff
@@ -37,6 +37,8 @@
 
 if TYPE_CHECKING:
     from rucio.daemons.common import HeartbeatHandler
+    from types import FrameType
+    from typing import Optional
```
Comment (dchristidis) [inline lib/rucio/daemons/atropos/atropos.py]: Please sort as you’ve done elsewhere (e.g. in `automatix.py`).

File: lib/rucio/daemons/c3po/c3po.py [hunk 3]
```diff
@@ -24,6 +24,7 @@
 from threading import Event, Thread
 from time import sleep
 from uuid import uuid4
+from typing import TYPE_CHECKING
```
Comment (dchristidis) [inline lib/rucio/daemons/c3po/c3po.py]: `typing` should come before `uuid`.

File: lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py [hunk 4]
```diff
@@ -27,6 +27,7 @@
 from sys import argv
 from typing import List, Any, Optional
 import re
+from typing import TYPE_CHECKING
```
Comment (dchristidis) [inline lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py]: There’s already a `from typing …` above.

File: lib/rucio/daemons/hermes/hermes.py [hunk 5]
```diff
@@ -50,7 +50,9 @@
 from rucio.daemons.common import run_daemon
 
 if TYPE_CHECKING:
-    from typing import Callable, List, Dict
+    from typing import Callable, List, Dict, Optional
+    from types import FrameType
```
Comment (dchristidis) [inline lib/rucio/daemons/hermes/hermes.py]: `types` should come before `typing`.

**Review by dchristidis — CHANGES_REQUESTED** [review]
This is likely to be my last request for changes. Please adjust some of the imports as noted in the comments below.

In addition, I would like to ask you to adjust the commit message. Please ensure that there is a blank line between the subject and the rest of the message, then move ‘Fixes #6075’ to the end of the subject.

**Comment by shalearkane** [comment]
Sir, could you please review the changes? I have made all the requested changes and have rebased the code against master branch.

**Review by dchristidis — APPROVED** [review]

**Pull request merged**